fvm/crypto/crypto.go

```diff
@@ -160,71 +164,83 @@ func VerifySignatureFromRuntime(
+		if crypto.IsInvalidInputsError(err) {
+			return false, err
+		}
+		panic(fmt.Errorf("verify runtime signature failed with unexpected error %w", err))
 	}
 
 	return valid, nil
 }
 
-type SignatureVerifier interface {
-	Verify(
-		signature []byte,
-		tag string,
-		message []byte,
-		publicKey crypto.PublicKey,
-		hashAlgo hash.HashingAlgorithm,
-	) (bool, error)
-}
-
-type DefaultSignatureVerifier struct{}
-
-func NewDefaultSignatureVerifier() DefaultSignatureVerifier {
-	return DefaultSignatureVerifier{}
-}
-
-func (DefaultSignatureVerifier) Verify(
+// VerifySignatureFromRuntime performs signature verification using raw values provided
+// by the Cadence runtime.
+//
+// The signature/hash function combinations accepted are:
+//   - ECDSA (on both curves P-256 and secp256k1) with any of SHA2-256/SHA3-256.
+// The tag is applied to the message as a constant length prefix.
+//
+// The function errors:
+//  - NewValueErrorf for any user error
+//  - panic for any other unexpected error
+func VerifySignatureFromTransaction(
 	signature []byte,
-	tag string,
 	message []byte,
-	publicKey crypto.PublicKey,
+	pk crypto.PublicKey,
 	hashAlgo hash.HashingAlgorithm,
 ) (bool, error) {
 
-	var hasher hash.Hasher
+	// check ECDSA compatibilites
+	if pk.Algorithm() != crypto.ECDSAP256 && pk.Algorithm() != crypto.ECDSASecp256k1 {
+		// TODO: check if we should panic
+		// This case only happens in production if there is a bug
+		return false, errors.NewValueErrorf(pk.Algorithm().String(), "is not supported in transactions")
+	}
+	// hashing compatibility
+	if hashAlgo != hash.SHA2_256 && hashAlgo != hash.SHA3_256 {
+		// TODO: check if we should panic
+		// This case only happens in production if there is a bug
+		return false, errors.NewValueErrorf(hashAlgo.String(), "is not supported in transactions")
+	}
 
-	switch hashAlgo {
-	case hash.SHA2_256, hash.SHA3_256, hash.Keccak_256:
-		var err error
-		if hasher, err = NewPrefixedHashing(hashAlgo, tag); err != nil {
-			return false, errors.NewValueErrorf(err.Error(), "verification failed")
-		}
-	case hash.KMAC128:
-		hasher = crypto.NewBLSKMAC(tag)
-	default:
-		return false, errors.NewValueErrorf(fmt.Sprint(hashAlgo), "hashing algorithm type not found")
+	hasher, err := NewPrefixedHashing(hashAlgo, flow.TransactionTagString)
+	if err != nil {
+		return false, errors.NewValueErrorf(err.Error(), "transaction verification failed")
 	}
 
-	valid, err := publicKey.Verify(signature, message, hasher)
+	valid, err := pk.Verify(signature, message, hasher)
 	if err != nil {
 		// All inputs are guaranteed to be valid at this stage.
 		// The check for crypto.InvalidInputs is only a sanity check
 		if crypto.IsInvalidInputsError(err) {
 			return false, err
 		}
-		panic(fmt.Errorf("verify signature failed with unexpected error %w", err))
+		// TODO: check if we should panic or bubble the error up
+		panic(fmt.Errorf("verify transaction signature failed with unexpected error %w", err))
```

Comment:
@ramtinms, wondering if we should panic or return an error here. This is a case of unexpected error (error in allocating memory for instance..)

Reply:
we usually panic if there is a small chance of non-deterministic reasons that caused this error.